File: SWENG/LCA/LCATree.go
package main

import (
	"fmt"
)

type Node struct{
	val int
	left *Node
	right *Node
}


func newNode(val int, left *Node, right *Node) *Node{
	return &Node{
		val: val,
		left : left,
		right : right,
	}
}

func LCA(root *Node, val1 int, val2 int) *Node{
	if root == nil {
		return newNode(-99999, nil,nil)
	}
	if search(root, val1) == false{
		return newNode(-99999, nil,nil)
	}
	if search(root, val2) == false{
		return newNode(-99999, nil,nil)
	}
	if search(root, val1) == true && search(root, val2) == true{
		return LowestCommonAncestor(root, val1, val2)
	}

	return nil
}

func search(root *Node, key int) bool{
	if root == nil{
		return false
	}
	if key < root.val {
		search(root.left, key)
	}
	if key > root.val{
		search(root.right, key)
	}
	return true
}

func LowestCommonAncestor(root *Node, val1 int, val2 int) *Node{
	if root == nil {
		return nil
	}

	if val1 > root.val && val2 > root.val{
		return LowestCommonAncestor(root.right, val1, val2)
	}

	if val1 < root.val && val2 < root.val {
		return LowestCommonAncestor(root.left, val1, val2)
	}

	return root
}

func main(){

	root := Node{100, nil, nil}
	root.left = &Node{50, nil, nil}
	root.right = &Node{150, nil, nil}
	root.left.left = &Node{20, nil, nil}
	root.left.left.left = &Node{10, nil, nil}
	root.left.left.right = &Node{30, nil, nil}
	root.left.right = &Node{80, nil, nil}
	root.right.left = &Node{110, nil, nil}
	root.right.right = &Node{200, nil, nil}

	var result = LowestCommonAncestor(&root, 10, 80)
	fmt.Println("LAC of 10 and 80 is", result.val)

	var result2 = LowestCommonAncestor(&root, 30, 200)
	fmt.Println("LAC of 30 and 200 is", result2.val)

	var result3 = LowestCommonAncestor(&root, 10, 30)
	fmt.Println("LCA of 10 and 30 is", result3.val)

}

File: SWENG/LCA/LCATree.java
package src;

class Node{

    int data;
    Node left, right;
    public Node(int value){
        data = value;
        left = right = null;
    }
}

public class LCATree {

    Node root;

    Node findLCA(int value1, int value2){
        if(search(root, value1) == false){
            return null;
        }
        if(search(root, value2) == false){
            return null;
        }
        if(search(root, value1) == true && search(root, value2) == true) {
            return findLCA(root, value1, value2);
        }
        else{
            return null;
        }
    }

    boolean search(Node root, int key){
        if(root == null){
            return false;
        }
        if(root.data == key){
            return true;
        }
        boolean result = (search(root.left, key));
        if(result){
            return true;
        }
        boolean result2 = search(root.right, key);
        return result2;
    }


    Node findLCA(Node node, int val1, int val2){
        // check for null node
        if (node == null) {
            return null;
        }

            //check if either input are the root node
            if (node.data == val1 || node.data == val2) {
                return node;
            }
            // Look for keys in left and right subtrees
            Node left_lca = findLCA(node.left, val1, val2);
            Node right_lca = findLCA(node.right, val1, val2);


            if (left_lca != null && right_lca != null) {
                return node;
            }

            return (left_lca != null) ? left_lca : right_lca;
        }

   /* public static void main(String[] args) {
        //create and fill tree
        LCATree tree = new LCATree();
        tree.root = new Node(1);
        tree.root.left = new Node(2);
        tree.root.right = new Node(3);
        tree.root.left.left = new Node(8);
        tree.root.left.right = new Node(9);
        tree.root.right.left = new Node(13);
        tree.root.right.right = new Node(15);

        System.out.println("LCA for 13 and 15 is " + tree.findLCA(13, 15).data);
        System.out.println("LCA for 8 and 9 is " + tree.findLCA(8,9).data);
        System.out.println("LCA for 2 and 15 is " + tree.findLCA(2,15).data);

    }*/
}
